storage_service/locations/models/wellcome.py

```diff
@@ -87,6 +87,120 @@ def mkdir_p(dirpath):
             raise
 
 
+def get_wellcome_identifier(src_path, default_identifier):
+    """
+    By default, Archivematica will use the UUID as the External-Identifier
+    when calling the Wellcome Storage.
+
+    This is somewhat unpleasant -- if you're browsing the storage without
+    Archivematica references, it's hard to know where to find a given archive.
+    For example, if you're looking for PPMIA/1/2, what UUID is that?
+
+    If all the objects in the bag have a common value in the dc.identifier field,
+    which should be a catalogue reference, use that in preference to the
+    Archivematica external identifier.
+
+    """
+    LOGGER.debug("Trying to find Wellcome identifier in %s", src_path)
+
+    # If we're not looking at a tar.gz compressed bag, stop.
+    if not src_path.endswith(".tar.gz"):
+        return default_identifier
+
+    # Unpack the tar.gz to a temporary directory.  We run tar in a subprocess
+    # because it's CPU intensive and we don't want to hang the main Archivematica
+    # thread.
+    temp_dir = tempfile.mkdtemp()
+    try:
+        subprocess.check_call(["tar", "-xzf", src_path, "-C", temp_dir])
+    except subprocess.CalledProcessError as err:
+        LOGGER.debug("Error uncompressing tar.gz bag: %r", err)
+        return default_identifier
+
+    # There should be a single directory in the temporary directory -- the
+    # uncompressed bag.
+    if len(os.listdir(temp_dir)) != 1:
+        LOGGER.debug(
+            "Unable to identify root of bag in: os.listdir(%r) = %r",
+            temp_dir, os.listdir(temp_dir)
+        )
+        return default_identifier
+
+    # Inside the bag, we look for the metadata.csv file that contains the
+    # dc.identifier values.  If we can't identify that file unambiguously,
+    # fall back to the UUID.
+    bag_dir = os.path.join(temp_dir, os.listdir(temp_dir)[0])
+    assert os.path.exists(bag_dir)
+    LOGGER.debug("Expanded bag into directory %s" % bag_dir)
+    bag = bagit.Bag(bag_dir)
+
+    manifest_csvs = [
+        name
+        for name in bag.payload_files()
+        if name.endswith("/metadata.csv")
```

Comment:
This check will fail if there's a random file in the digital archive called `metadata.csv`.

We can say exactly which it is – in a path of a particular format and depth.